### PR TITLE
ios/android: allow zero for total upstream timeouts

### PR DESCRIPTION
**Pull request by rebello95**
We should allow zero to be specified for `totalUpstreamTimeoutMS`, which effectively [disables the upstream timeout per Envoy's docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#envoy-api-field-route-routeaction-timeout).

This can be useful for consumers who only want to timeout on a per-request basis (rather than timing out all retries in aggregate).

Signed-off-by: Michael Rebello <me@michaelrebello.com>

### CODE REVIEW

**Comment by rebello95** [comment]
> aren't we worried that API for perRetryTimeoutMS and totalUpstreamTimeoutMS are different? One of them accepts nil while the other relies on 0 being passed. Could we come up with something that would work for both of them - accepting nils for both of them or 0 for both of them?

This is a good call out. I'll update both to be nullable.

**Comment by rebello95** [comment]
For context, there's a difference in how those are serialized on the wire. For upstream/route timeouts, we need to manually specify `0` in the header, whereas for request timeouts, we omit the header. But I don't think we should expose these implementation details in the public platform interface.

File: library/swift/src/RetryPolicy.swift [hunk 1]
```diff
@@ -47,13 +47,15 @@ public final class RetryPolicy: NSObject {
...
-              totalUpstreamTimeoutMS: UInt = 15_000)
+              totalUpstreamTimeoutMS: UInt? = 15_000)
   {
-    if let perRetryTimeoutMS = perRetryTimeoutMS {
-      assert(perRetryTimeoutMS <= totalUpstreamTimeoutMS || totalUpstreamTimeoutMS == 0,
-             "Per-retry timeout must be <= total timeout if total timeout != 0")
+    if let perRetryTimeoutMS = perRetryTimeoutMS,
+      let totalUpstreamTimeoutMS = totalUpstreamTimeoutMS
+    {
+      assert(perRetryTimeoutMS <= totalUpstreamTimeoutMS,
```
Comment (Augustyniak) [inline library/swift/src/RetryPolicy.swift]: you still need to check for `totalUpstreamTimeoutMS ` here, right?

Comment (Augustyniak) [reply library/swift/src/RetryPolicy.swift]: assert(perRetryTimeoutMS <= totalUpstreamTimeoutMS && totalUpstreamTimeoutMS != 0)

File: library/swift/test/RetryPolicyMapperTests.swift [hunk 2]
```diff
@@ -23,4 +23,9 @@ final class RetryPolicyMapperTests: XCTestCase {
     let policy = RetryPolicy(maxRetryCount: 123, retryOn: RetryRule.allCases)
     XCTAssertNil(policy.outboundHeaders()["x-envoy-upstream-rq-per-try-timeout-ms"])
   }
+
+  func testConvertingToHeadersWithoutUpstreamTimeoutIncludesZeroForTimeoutHeader() {
```
Comment (Augustyniak) [inline library/swift/test/RetryPolicyMapperTests.swift]: let's add a test case for when per request timeout is equal to 0 (not nil) and per retry timeout is greater than 0.

Comment (rebello95) [reply library/swift/test/RetryPolicyMapperTests.swift]: Considered that but iOS doesn't really provide a way to test that `assert`s fail, and making the builder `throw` doesn't really seem worth it here

File: library/swift/src/RetryPolicy.swift (hunk 1 above)
Comment (rebello95) [reply library/swift/src/RetryPolicy.swift]: Personally I think that omitting the `totalUpstreamTimeoutMS != 0` exclusion from the assert is probably preferred here since the consumer will then change the `0` to `nil` to achieve the same behavior (in line with the documentation of this function). Thoughts?

Comment (Augustyniak) [reply library/swift/src/RetryPolicy.swift]: One question - What happens if I provide `perRetryTimeoutMS = 0` and `totalUpstreamTimeoutMS = 0`? I would expect this too break since the docs says that a user should use `nil` if they want `totalUpstreamTimeoutMS` to be ignored.

Comment (rebello95) [reply library/swift/src/RetryPolicy.swift]: That's fair, I can update to include it in the assert.
